int32 id
bool mask